File: python/fretboard.py
import notes

STRINGS = ['E', 'A', 'D', 'G', 'B', 'E']
FRET_COUNT = 12

# Coordinate constants for mapping fret positions to image pixels
x_i, y_i = 45, 57
x_bar, y_bar = 120, 45


def get_offset(fret: int) -> int:
    ''' Returns the pixel offset for a given fret
        because frets are not equal width. '''
    offsets = {0: 0, 1: 46, 2: 39, 3: 39, 4: 36, 5: 36,
               6: 40, 7: 50, 8: 57, 9: 72, 10: 85, 11: 105, 12: 125}
    return offsets.get(fret, 0)


def order_notes(desired_notes: list, strings: list = STRINGS) -> list:
    ''' Creates a nested list of all desired_notes on the fretboard. '''
    return [notes.order_notes(desired_notes, string, FRET_COUNT)
            for string in reversed(strings)]


def get_positions(fretboard: list, note: str) -> dict:
    ''' Returns {note: [(string_idx, fret_idx), ...]} for a single note. '''
    positions = [
        (string_idx, fret_idx)
        for string_idx, string in enumerate(fretboard)
        for fret_idx, n in enumerate(string)
        if n == note
    ]
    return {note: positions}


def get_notes(desired_notes: list, fretboard: list) -> dict:
    ''' Returns {note: [(string_idx, fret_idx), ...]}
        for all desired_notes.
    '''
    result = {}
    for note in desired_notes:
        result |= get_positions(fretboard, note)
    return result


def map_notes(desired_notes: list, fretboard_notes: dict) -> dict:
    ''' Maps each note to pixel coordinates on the fretboard image. '''
    fretboard_map = {}
    for note in desired_notes:
        fret_positions = fretboard_notes[note]
        positions = []
        for string_idx, fret_idx in fret_positions:
            y_pos = y_i + (string_idx * y_bar)
            x_pos = x_i + (fret_idx * x_bar) - get_offset(fret_idx)
            positions.append((x_pos, y_pos))
        fretboard_map[note] = positions
    return fretboard_map


def create(desired_notes: list) -> dict:
    ''' Returns the completed fretboard coordinate map. '''
    fretboard_grid = order_notes(desired_notes)
    fretboard_notes = get_notes(desired_notes, fretboard_grid)
    return map_notes(desired_notes, fretboard_notes)

File: web/src/lib/fretboard.ts
import { orderNotes as notesOrderNotes } from './notes';

export const STRINGS = ['E', 'A', 'D', 'G', 'B', 'E'] as const;
export const FRET_COUNT = 12;

// Coordinate constants for mapping fret positions to image pixels
const X_I = 32;
const Y_I = 42;
const X_BAR = 120;
const Y_BAR = 45;

const FRET_OFFSETS: Record<number, number> = {
  0: 0, 1: 46, 2: 39, 3: 39, 4: 36, 5: 36,
  6: 40, 7: 50, 8: 57, 9: 72, 10: 85, 11: 105, 12: 125,
};

export function getOffset(fret: number): number {
  return FRET_OFFSETS[fret] ?? 0;
}

export type FretboardGrid = string[][];
export type NotePositions = Record<string, [number, number][]>;

export function buildGrid(desiredNotes: string[]): FretboardGrid {
  return [...STRINGS].reverse().map((string) =>
    notesOrderNotes(desiredNotes, string, FRET_COUNT)
  );
}

function getPositions(grid: FretboardGrid, note: string): [number, number][] {
  const positions: [number, number][] = [];
  for (let stringIdx = 0; stringIdx < grid.length; stringIdx++) {
    for (let fretIdx = 0; fretIdx < grid[stringIdx].length; fretIdx++) {
      if (grid[stringIdx][fretIdx] === note) {
        positions.push([stringIdx, fretIdx]);
      }
    }
  }
  return positions;
}

export function getNotes(desiredNotes: string[], grid: FretboardGrid): Record<string, [number, number][]> {
  return Object.fromEntries(desiredNotes.map((note) => [note, getPositions(grid, note)]));
}

export function mapNotes(desiredNotes: string[], fretboardNotes: Record<string, [number, number][]>): NotePositions {
  const result: NotePositions = {};
  for (const note of desiredNotes) {
    result[note] = fretboardNotes[note].map(([stringIdx, fretIdx]) => [
      X_I + fretIdx * X_BAR - getOffset(fretIdx),
      Y_I + stringIdx * Y_BAR,
    ]);
  }
  return result;
}

export function createFretboard(desiredNotes: string[]): NotePositions {
  const grid = buildGrid(desiredNotes);
  const fretboardNotes = getNotes(desiredNotes, grid);
  return mapNotes(desiredNotes, fretboardNotes);
}
